src/govuk/components/header/header.mjs

```diff
@@ -8,34 +8,68 @@ function Header ($module) {
   this.$menu = this.$menuButton && $module.querySelector(
     '#' + this.$menuButton.getAttribute('aria-controls')
   )
+
+  // Save the opened/closed state for the nav in memory so that we can
+  // accurately maintain state when the screen is changed from small to
+  // big and back to small
+  this.menuIsOpen = false
+
+  // A global const for storing a matchMedia instance which we'll use to
+  // detect when a screen size change happens. We set this later during the
+  // init function and rely on it being null if the feature isn't available
+  // to initially apply hidden attributes
+  this.mql = null
 }
 
 /**
  * Initialise header
  *
  * Check for the presence of the header, menu and menu button – if any are
  * missing then there's nothing to do so return early.
+ * Feature sniff for and apply a matchMedia for desktop which will
+ * trigger a state sync if the browser viewport moves between states. If
+ * matchMedia isn't available, hide the menu button and present the "no js"
+ * version of the menu to the user.
  */
 Header.prototype.init = function () {
   if (!this.$module || !this.$menuButton || !this.$menu) {
     return
   }
 
-  this.syncState(this.$menu.classList.contains('govuk-header__navigation-list--open'))
-  this.$menuButton.addEventListener('click', this.handleMenuButtonClick.bind(this))
+  if ('matchMedia' in window) {
+    // Set the matchMedia to the govuk-frontend desktop breakpoint
+    this.mql = window.matchMedia('(min-width: 48.0625em)')
+    this.mql.addEventListener('change', this.syncState.bind(this))
```

Comment:
Unfortunately it looks like [IE11 and older versions of Safari (< 14.0)](https://caniuse.com/mdn-api_mediaquerylist) don't support `MediaQueryList.addEventListener` even though they support `window.matchMedia`, so at the minute this doesn't work in those browsers (and throws errors).

We need to use the non-standard `MediaQueryList.addListener` instead, [like we do for tabs](https://github.com/alphagov/govuk-frontend/blob/159455d6c049f99242569afc87302337308ee8ff/src/govuk/components/tabs/tabs.mjs#L24-L28):

```suggestion
    this.mql.addListener(this.syncState.bind(this))
```

Or, if we wanted to future-proof the code against the deprecation:

```suggestion
    if (this.mql.hasOwnProperty('addEventListener')) {
      this.mql.addEventListener('change', this.syncState.bind(this))
    } else {
      this.mql.addListener(this.syncState.bind(this))
    }
```

Either way, the `addListener` method is deprecated (and this gets flagged by my editor), so might also be worth adding a comment to explain what's going on?

Reply:
I amended the solution very slightly as js apparently doesn't like `this.mql.hasOwnProperty('addEventListener')` but this should now be working 👍🏻

Reply:
After a chat on slack, this is now using `'addEventListener in this.mql` as it's tidier and more consistent with what I'm doing to sniff for `matchMedia`. The reason the codebase didn't like `hasOwnProperty` was due to a linting concern.